### PR TITLE
feature: allow "cycle properties", like "theories[-1]" in SimpleCycle

### LINKED ISSUE #121: Get a real Experimentalist working in the closed loop

### PR DESCRIPTION
**Pull request by hollandjg**
## Description

Add an easy way to pass variables like "the last theory" and "all the data up until now" into the `_SimpleCycle` closed loop runner.

resolves #121 
related to #154 

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- `_SimpleCycle` Cycle-implementation extended to allow automatically updating variables, like "all the independent variable values up until now" and "the last theorist" to be made available to an experimentalist.
- Added an example notebook which uses the new machinery when running the popper experimentalist.

Example code:
```python
poppernet_experimentalist = make_pipeline(
    [poppernet_pooler, nearest_values_sampler],
)

poppernet_experimentalist_cycle = Cycle(
    metadata=study_metadata,
    theorist=bms_theorist,
    experimentalist=poppernet_experimentalist,
    experiment_runner=example_synthetic_experiment_runner,
    params={"experimentalist" : {
                "poppernet_pooler": {
                    "model": "%theories[-1]%", # 👈 magic variable here,
                    "x_train": "%observations.ivs%", # 👈 ... and here,
                    "y_train": "%observations.dvs%", # 👈 ... and here.
                    "metadata": study_metadata,
                    "num_samples": observations_per_cycle,
                },
                "nearest_values_sampler": {
                    "allowed_values": study_metadata.independent_variables[0].allowed_values
                }
            }
        }
    )
```

### CODE REVIEW

File: autora/cycle/_simple_cycle.py [hunk 1]
```diff
@@ -267,3 +389,97 @@ def _theorist_callback(theorist, data_in: _SimpleCycleData):
...
+
+    Args:
+        params: a (nested) dictionary of keys and values, where some values might be "magic"
+        magics: a dictionary of "magic values" (keys) and "real values" (values)
+
+    Returns: a (nested) dictionary where "magic values" are replaced by the "real values"
+
+    Examples:
+
+        >>> params_0 = {"key": "%magic_number%"}
```
Comment (musslick) [inline autora/cycle/_simple_cycle.py:409]: I'm wondering: What is the downside to passing a callable instead?

**Review by musslick — CHANGES_REQUESTED** [review]
This looks like an incredibly powerful feature to be added! I had some trouble understanding whats going on and most suggestions are minor typos. It might be worth revisiting how to write the function documentation, so that's easier to parse how to use the magic variables–Perhaps we can figure out a more minimal example?

File: autora/experimentalist/sampler/__init__.py [hunk 2]
```diff
@@ -1,2 +1,3 @@
+from .poppernet import nearest_values_sampler, poppernet_pooler
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/experimentalist/sampler/__init__.py:1]: Yes, we'll have to update it after 180 is merged.

File: autora/cycle/_simple_cycle.py [hunk 3]
```diff
@@ -173,7 +174,106 @@ class _SimpleCycle:
         Generated 111 theories
 
 
+        It's easy to pass parameters to the cycle components, if there are any needed.
+        Here we have an experimentalist which takes a parameter:
+        >>> uniform_random_rng = np.random.default_rng(180)
+        >>> def uniform_random_sampler(num_samples):
+        ...     return uniform_random_rng.uniform(low=0, high=11, size=num_samples)
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:181]: - [ ] change all `num_samples` to `n`

File: autora/cycle/_simple_cycle.py [hunk 4]
```diff
@@ -173,7 +174,106 @@ class _SimpleCycle:
...
+        ...     ]
+        ... )
+        >>> magics_cycle = _SimpleCycle(
+        ...     metadata=magics_metadata,
+        ...     theorist=example_theorist,
+        ...     experimentalist=unobserved_data_experimentalist,
+        ...     experiment_runner=example_synthetic_experiment_runner,
+        ...     params={
+        ...         "experimentalist": {
+        ...             "exclude_conditions": {"excluded_conditions": "%observations.ivs%"},
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:233]: - [ ] Need documentation on which magic parameters exist

[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:233]: Perhaps rename – "lambda key" – what exactly does "lambda" mean

[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:233]: - [ ] Is it easier just to inject a lambda function?

File: autora/cycle/_simple_cycle.py [hunk 5]
```diff
@@ -173,7 +174,106 @@ class _SimpleCycle:
...
+        >>> _ = cycle_with_parameters.run()
+        >>> cycle_with_parameters.data.conditions[-1].flatten()
+        array([10.5838232 ,  9.45666031])
+
+
+         In the case we have an experimentalist which needs variables such as
+        - the current best theory
+        - all the existing observational data
+        to generate its next values, we can use magic parameters:
+        >>> magics_metadata = VariableCollection(
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py]: Relabelled it as discussed

File: autora/cycle/_simple_cycle.py [hunk 6]
```diff
@@ -267,3 +389,97 @@ def _theorist_callback(theorist, data_in: _SimpleCycleData):
     def _monitor_callback(self, data: _SimpleCycleData):
         if self.monitor is not None:
             self.monitor(data)
+
+
+def _resolve_magic_params(params: Dict, magics: Mapping):
+    """
+    Resolve "magic values" in a nested dictionary.
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py]: updated as discussed

File: autora/cycle/_simple_cycle.py (hunk 1 above)
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py]: let's make a new issue for that #189

File: autora/cycle/_simple_cycle.py [hunk 7]
```diff
@@ -173,7 +174,106 @@ class _SimpleCycle:
         Generated 111 theories
 
 
+        It's easy to pass parameters to the cycle components, if there are any needed.
+        Here we have an experimentalist which takes a parameter:
+        >>> uniform_random_rng = np.random.default_rng(180)
+        >>> def uniform_random_sampler(num_samples):
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py]: done

File: autora/cycle/_simple_cycle.py [hunk 8]
```diff
@@ -1,18 +1,19 @@
...
+from autora.utils.dictionary import LazyDict
 from autora.variable import VariableCollection
 
 
 @dataclass(frozen=True)
-class _SimpleCycleData:
-    """An object passed between processing steps in the _SimpleCycle which holds all the
-    data which can be updated."""
+class SimpleCycleData:
+    """An object passed between and updated by  processing steps in the SimpleCycle."""
```
Comment (musslick) [inline autora/cycle/_simple_cycle.py:16]:
```suggestion
    """An object passed between and updated by processing steps in the SimpleCycle."""
```

**Review by musslick — APPROVED** [review]
All changes look great to me; can't wait to try it out. I only had a very minor commit suggestions for a double-space in the comments...

File: autora/cycle/_simple_cycle.py [hunk 9]
```diff
@@ -26,26 +27,85 @@ class _SimpleCycleData:
...
+        '%observations.dvs%', '%theories[-1]%', '%theories%']
+
+    """
+
+    n_ivs = len(data.metadata.independent_variables)
+    n_dvs = len(data.metadata.dependent_variables)
+    cycle_property_dict = LazyDict(
+        {
+            "%observations.ivs[-1]%": lambda: data.observations[-1][:, 0:n_ivs],
+            "%observations.dvs[-1]%": lambda: data.observations[-1][:, n_ivs:],
```
Comment (benwandrew) [inline autora/cycle/_simple_cycle.py:55]: guessing i'm not parsing this quite right, but just wanted to confirm: should be `n_ivs:` and not `n_dvs:`? 

same question with line 59 below, too.

File: autora/cycle/_simple_cycle.py [hunk 10]
```diff
@@ -136,15 +196,15 @@ class _SimpleCycle:
...
-        <_simple_cycle._SimpleCycle object at 0x...>
+        <simple.SimpleCycle object at 0x...>
 
         >>> next(cycle) # doctest: +ELLIPSIS
         Generated 6 theories
-        <_simple_cycle._SimpleCycle object at 0x...>
+        <simple.SimpleCycle object at 0x...>
 
         We can continue to run the cycle as long as we like,
         with a simple arbitrary stopping condition like the number of theories generated:
```
Comment (benwandrew) [inline autora/cycle/_simple_cycle.py:210]: not sure how to suggest a change to an unmodified line, but the second 'between' in the below should be 'of'. ie, '...difference between the gradients of...'

`... or the precision (here we keep iterating while the difference between the gradients
        between the second-last and last cycle is larger than 1x10^-3).`

**Review by benwandrew — CHANGES_REQUESTED** [review]
couple trivial text suggestions and a quick confirmation in the initialization of `cycle_property_dict`

File: autora/cycle/_simple_cycle.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:55]: Yes, this is correct, I think. 

- The first `n_ivs` columns are what you want for the IVs.
- Then you want the rest, which you could write more verbosely as `data.observations[-1][:, n_ivs : n_ivs+n_dvs]`

The first DV is at the column number given by n_ivs, if we stack the IVs and then the DVs.

In this example, we have 3 IVs: `x0`, `x1`, and `x2` and 2 DVs: `y0` and `y1`
```
columns: 
    column index: |  0 |  1 |  2 |  3 |  4 |
    label:        | x0 | x1 | x2 | y0 | y1 |

n_ivs = 3
n_dvs = 2
```

Rules when doing a column slice:

- The first column you specify is _included_ 
- If unspecified, the first column is assumed to be the first column in the table, `0`
- The last column you specify is _excluded_
- If unspecified, the last column is assumed to be the (last + 1)th column in the table, `len(columns)`, which in this case is `5`.

For the IVs, you want columns 0, 1 and 2, which you get by 
```
   columns[0:3] 
== columns[0:n_ivs]
== columns[:n_ivs]
```

For the DVs, you want columns 3 and 4, which you get by 
```  
   columns[3:5] 
== columns [n_ivs:n_ivs+n_dvs] 
== columns [n_ivs:]
```

File: autora/cycle/_simple_cycle.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hollandjg) [reply autora/cycle/_simple_cycle.py:210]: Thanks! Done

**Review by benwandrew — APPROVED** [review]
nice work, this is a juicy one!

**Pull request merged**